File: src/main/scala/io/corbel/sdk/iam/Iam.scala
package io.corbel.sdk.iam

import io.corbel.sdk.api.{RequestParams, Query}
import io.corbel.sdk.auth.AuthenticationProvider._
import io.corbel.sdk.error.ApiError

import scala.concurrent.{ExecutionContext, Future}

/**
  * @author Alexander De Leon (alex.deleon@devialab.com)
  */
trait Iam extends AuthenticationClient {

  /* ----------------- Scopes ---------------------- */

  def getScope(id: String)(implicit authenticationProvider: AuthenticationProvider = null, ec: ExecutionContext): Future[Either[ApiError,Scope]]

  /* ----------------- Users ---------------------- */

  def getUserById(id: String)(implicit authenticationProvider: AuthenticationProvider = null, ec: ExecutionContext): Future[Either[ApiError,User]]

  def getUser(implicit authenticationProvider: AuthenticationProvider = null, ec: ExecutionContext): Future[Either[ApiError,User]]

  def addGroupsToUser(userId: String, groups: Iterable[String])(implicit authenticationProvider: AuthenticationProvider = null, ec: ExecutionContext): Future[Either[ApiError,Unit]]

  def updateUser(user: User)(implicit authenticationProvider: AuthenticationProvider = null, ec: ExecutionContext): Future[Either[ApiError,Unit]]

  def findUsers(params: RequestParams)(implicit authenticationProvider: AuthenticationProvider = null, ec: ExecutionContext): Future[Either[ApiError,Seq[User]]]

  /* ----------------- Groups ---------------------- */

  def createGroup(group: Group)(implicit authenticationProvider: AuthenticationProvider = null, ec: ExecutionContext): Future[Either[ApiError,String]]
}

File: src/main/java/io/corbel/sdk/java/iam/Iam.java
package io.corbel.sdk.java.iam;

import io.corbel.sdk.api.RequestParams;
import io.corbel.sdk.error.ApiError;
import io.corbel.sdk.iam.*;
import scala.Unit;
import scala.util.Either;

import java.util.Collection;
import java.util.List;
import java.util.Optional;
import java.util.concurrent.CompletionStage;

/**
 * @author Alexander De Leon (alex.deleon@devialab.com)
 */
public interface Iam {

    /* ----------------- Authentication ------------ */

    CompletionStage<Either<ApiError, AuthenticationResponse>> authenticate(ClientCredentials clientCredentials,
                                                                           Optional<UserCredentials> userCredentials,
                                                                           AuthenticationOptions authenticationOptions);

    CompletionStage<Either<ApiError, AuthenticationResponse>> authenticationRefresh(ClientCredentials clientCredentials,
                                                                           String refreshToken,
                                                                           AuthenticationOptions authenticationOptions);

  /* ----------------- Scopes ---------------------- */

    CompletionStage<Either<ApiError, Scope>> getScope(String id);


  /* ----------------- Users ---------------------- */

    CompletionStage<Either<ApiError, String>> createUser(User user);

    CompletionStage<Either<ApiError, User>> getUserById(String id);

    CompletionStage<Either<ApiError, User>> getUserIdByUsername(String username);

    CompletionStage<Either<ApiError, User>> getUser();

    CompletionStage<Either<ApiError, Unit>> updateUser(User user);

    CompletionStage<Either<ApiError, Collection<Device>>> getUserDevices(String userId);

    CompletionStage<Either<ApiError, Unit>> addGroupsToUser(String userId, Collection<String> groups);

    CompletionStage<Either<ApiError, Unit>> deleteGroupToUser(String userId, String groupId);

    CompletionStage<Either<ApiError, Collection<User>>> findUsers(RequestParams params);


  /* ----------------- Groups ---------------------- */

    CompletionStage<Either<ApiError, String>> createGroup(Group group);

}
